ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,467 @@
+
+## Specification
+
+* Unicode version `15.0.0`
+	* Normalization is a living specification and should use the latest stable version of Unicode.
+* [`spec.json`](./ensip-15/spec.json) contains all [necessary data](#description-of-specjson) for normalization.
+* [`nf.json`](./ensip-15/nf.json) contains all necessary data for [Unicode Normalization Forms](https://unicode.org/reports/tr15/) NFC and NFD.
+
+### Definitions
+
+* Terms in **bold** throughout this document correspond with [components of `spec.json`](#description-of-specjson).
+* A string is a sequence of Unicode codepoints.
+	* Example: `"abc"` is `61 62 63`
+* An [Unicode emoji](https://www.unicode.org/reports/tr51/) is a [single entity](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries) composed of one or more codepoints:
+	* An **Emoji Sequence** is the preferred form of an emoji, resulting from input that [tokenized](#tokenize) into an `Emoji` token.
+		* Example: `💩︎︎ [1F4A9]` → `Emoji[1F4A9 FE0F]`
+			* `1F4A9 FE0F` is the **Emoji Sequence**.
+	* [`spec.json`](#description-of-specjson) contains the complete [list of valid](./ensip-15/emoji.md) **Emoji Sequences**.		
+		* [Derivation](#derivation) defines which emoji are normalizable.
+		* Not all Unicode emoji are valid.
+			* `‼ [203C] double exclamation mark` → *error: Disallowed character*
+			* `🈁 [1F201] Japanese “here” button` → `Text["ココ"]`
+	* An **Emoji Sequence** may contain characters that are disallowed:
+		* `👩‍❤️‍👨 [1F469 200D 2764 FE0F 200D 1F468] couple with heart: woman, man` — contains ZWJ
+		* `#️⃣ [23 FE0F 20E3] keycap: #` — contains `23 (#)`
+		* `🏴󠁧󠁢󠁥󠁮󠁧󠁿 [1F3F4 E0067 E0062 E0065 E006E E0067 E007F]` — contains `E00XX`
+	* An **Emoji Sequence** may contain other emoji:
+		* Example: `❤️ [2764 FE0F] red heart` is a substring of `❤️‍🔥 [2764 FE0F 200D 1F525] heart on fire` 
+	* Single-codepoint emoji may have various [presentation styles](https://www.unicode.org/reports/tr51/#Presentation_Style) on input:
+		* Default: `❤ [2764]`
+		* Text: `❤︎ [2764 FE0E]`
+		* Emoji: `❤️ [2764 FE0F]`
+	* However, these all [tokenize](#tokenize) to the same **Emoji Sequence**.
+	* All **Emoji Sequence** have explicit emoji-presentation.
+	* The convention of ignoring presentation is difficult to change because:
+		* Presentation characters (`FE0F` and `FE0E`) are **Ignored**
+	 	* [ENSIP-1](./ensip-1-ens.md) did not treat emoji differently from text
+		* Registration hashes are immutable
+	* [Beautification](#annex-beautification) can be used to restore emoji-presentation in normalized names.
+	
+### Algorithm
+
+* Normalization is the process of canonicalizing a name before for [hashing](./ensip-1-ens.md#namehash-algorithm).
+* It is idempotent: applying normalization multiple times produces the same result.
+* For user convenience, leading and trailing whitespace should be trimmed before normalization, as all whitespace codepoints are disallowed.  Inner characters should remain unmodified.

```

Comment:
I'm not sure I follow this one, can you suggest what the remedy would be?  I don't see how "leading and trailing whitespace" can be conflated with names/labels.

Reply:
Sorry, take or leave this suggestion if it isn't clear.

> For user convenience, **leading and trailing whitespace should be trimmed** before normalization, as all whitespace codepoints are disallowed.  Inner characters should remain unmodified.

From my point of view, listing this bullet under algorithm makes me think it's part of the normalization algorithm to strip any leading and trailing whitespace a user may have inputted for the user's convenience. I originally implemented this and it led to failing negative-case tests because they were looking for errors related to this. Once I realized this is meant from the *user's* perspective, I stopped trimming leading and trailing whitespace in our implementation and the tests failed with the expected failures.